testing/test_warnings.py

```diff
@@ -641,3 +642,151 @@ def pytest_configure():
     assert "INTERNALERROR" not in result.stderr.str()
     warning = recwarn.pop()
     assert str(warning.message) == "from pytest_configure"
+
+
+class TestStackLevel:
+    @pytest.fixture
+    def capwarn(self):
```

Comment:
This is a good approach, but I think we can improve it a bit:

1. Instead of only holding the last warning, it would be better to accumulate them on a list. This way we ensure the tests will fail with better messages in case more than one warning is raised in the future.

2. Following 1, you can return the list object directly, and automatically register it in `testdir`:

   ```python
       
       @pytest.fixture
       def capwarn(self, testdir):
           class CapturedWarnings:
               captured = []
   
               @classmethod
               def pytest_warning_captured(cls, warning_message, when, item, context):
                   cls.captured.append((warning_message, context))
   
           testdir.plugins = [CapturedWarnings()]
   
           return CapturedWarnings.captured
  ```

Reply:
To accumulate them makes sense to me but i am concerned about setting testdir.plugins in that fixture. During writing the tests i discovered that it is (strangely) not uniform to apply a plugin to pytests internal tests. While it works most of the time it **does not** work the above way in the [cacheprovider test](https://github.com/phloose/pytest/blob/acceptance-tests_pytest-dev/pytest/%234445/testing/test_warnings.py#L760) (if i am missing something please point me in the right direction). Therefore i would propose to still use the class as return value, so that you can do both ways.

Reply:
Done!